org.adempiere.base/src/org/adempiere/base/AnnotationBasedModelFactory.java

```diff
@@ -0,0 +1,65 @@
+package org.adempiere.base;
+
+import org.atteo.classindex.ClassIndex;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.wiring.BundleWiring;
+import org.osgi.service.component.ComponentContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ServiceScope;
+
+/**
+ * Translates table names into model classes having the {@link Model} annotation. Relies on
+ * {@link DefaultModelFactory} for everything else.
+ * This factory is designed to have a service rank higher than {@link DefaultModelFactory}, as class
+ * discovery using SPI is preferred over reflection-based methods.
+ * @author Saulo Gil
+ */
+@Component(scope = ServiceScope.BUNDLE)
+public class AnnotationBasedModelFactory extends DefaultModelFactory implements IModelFactory
+{
+
+	private Bundle usingBundle;
+
+	@Activate
+	void activate(ComponentContext context)
+	{
+		this.usingBundle = context.getUsingBundle();
+	}
+
+	@Override
+	public Class<?> getClass(String tableName) 
+	{
+		// search first into DefaultModelFactory's cache
+		Class<?> candidate = s_classCache.get(tableName);
+		if (candidate != null)
+		{
+			// Object.class indicate no generated PO class for tableName
+			if (candidate.equals(Object.class))
+				return null;
+			else
+				return candidate;
+		}
+
+		// scan annotations
+		BundleWiring wiring = usingBundle.adapt(BundleWiring.class);
+		for(Class<?> clazz : ClassIndex.getAnnotated(Model.class, wiring.getClassLoader()))
+		{
+			Model ma = clazz.getAnnotation(Model.class);
+			if(ma.table().equalsIgnoreCase(tableName)) 
+			{
+				candidate = clazz;
+				// M* classes are preferred over X* classes
+				if(!ma.intermediate())
```

Comment:
Don't fancy we take the intermediate class here. With this, if we don't manually annotate each and every existing M* classes, the X* will bed used instead. As I've posted below, I think we need to have facility to generate or modify M* classes instead.

Reply:
Just come across this http://www.atteo.org/static/classindex/apidocs/org/atteo/classindex/ClassIndex.html#getSubclasses-java.lang.Class- and I'm wondering whether we can use this to get rid of the need to use intermediate and the need to annotate M* classes ?

Reply:
Excellent idea, implemented at 9d9a22eb749f5075bdfebd1aecd20ea4158daf4b.